### ISSUE TITLE
Snowflake lambda expression trigger RF02 & RF03

### ISSUE THREAD

**Issue by Orsenna187** [sqlfluff/sqlfluff#6683]
### Search before asking

- [x] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

The higher-order functions `TRANSFORM()`, `REDUCE()` and `FILTER()` in Snowflake require the use of a lambda expressions. However, the syntax below wrongly triggers RF02 & RF03 with the anonymous <a> being flagged as an unqualified reference. 

`SELECT TRANSFORM(tbl.array, a INT -> a * 2)
FROM table AS tbl`

### Expected Behaviour

`SELECT TRANSFORM(tbl.array, a INT -> a * 2)
FROM table AS tbl`

This is valid SQL which shouldn't trigger rules RF02 and RF03.

### Observed Behaviour

`SELECT TRANSFORM(tbl.array, a INT -> a * 2)
FROM table AS tbl`

This is triggers rules RF02 & RF03.

Note that there is no workaround since the syntax below returns as syntax error.

`SELECT TRANSFORM(tbl.array, tbl.a INT -> tbl.a * 2)
FROM table AS tbl`

### How to reproduce

Simply use sqlfluff on any Snowflake sql using a higher-order function with a lambda expression. 

### Dialect

Snowflake

### Version

3.2.5

### Configuration

```
[sqlfluff]
# The dialect of SQL you're using (e.g., postgres, mysql, snowflake)
dialect = snowflake

# Linter behavior
exclude_rules = ST02,ST06,ST07

# Formatting settings
indent_unit = space 
tab_space_size = 4         
max_line_length = 120

[sqlfluff:rules:capitalisation.keywords]
capitalisation_policy = upper  

[sqlfluff:rules:capitalisation.identifiers]
capitalisation_policy = lower  

[sqlfluff:templater:jinja]
load_macros_from_path = dbt/macros

[sqlfluff:templater:jinja:context]
target={"name":"prod"}
```

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [x] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by shervinmathieu** [comment]
This rule specifically fails when there are multiple tables in the statement (or a table alias), and is not specific to Snowflake, other dialects will trigger the rule as well.

The following block will trigger RF02 on all dialects (function names do not matter) :

```
select
  filter(ref.ident, i -> i:value > 0) as sample_filter,
  transform(ref.ident, j -> j:value) as sample_transform
from ref
inner join ref2
  on ref.id = ref2.id;
```

**Comment by shervinmathieu** [comment]
Culprit seems to be the `_get_lambda_argument_columns` function in select parsing which is not correctly adding lambda identifiers to the `standalone_aliases` array.
 https://github.com/sqlfluff/sqlfluff/blob/e1448f8e7f459195a77b03cbb7f9010058c429e0/src/sqlfluff/utils/analysis/select.py#L212

**Comment by shervinmathieu** [comment]
This seems to be a little bit more complicated than I first thought.

Let's say that there is a table A with a column col1 and an array arr1.
This is valid SQL and the lambda expression does not reference col1 and instead transfroms arr1 :
`select transform(arr1, col1-> col1 + 1) as t from A`

Adding the aliases to the `standalone_aliases` array so that sqlfluff ignores them would in this case break rule checking for the col1 column.

Instead, it looks like rules RF02 and RF03 need to be patched to behave differently when the token tree is inside a `lambda_expression` block.

**Comment by shervinmathieu** [comment]
Ignore my previous comment, in the end it was the `_get_lambda_argument_columns` function that needed Snowflake support.
Opened a PR to fix this.

**Comment by Orsenna187** [comment]
Thanks for the quick PR 🙌
